src/checkers/inference/model/serialization/ToStringSerializer.java

```diff
@@ -24,49 +28,75 @@
 import checkers.inference.model.VariableSlot;
 
 /**
- * This Serializer is meant only to convert constraints and variables to
- * human readable strings.  It is used currently by the DebugSolver.
+ * This Serializer is meant only to convert constraints and variables to human readable strings. It
+ * is used currently by the DebugSolver.
  */
 public class ToStringSerializer implements Serializer<String, String> {
     private final boolean showAstPaths;
     private boolean showVerboseVars;
 
-    private int indent = 0;
-    public static final String INDENT_STRING = "    ";
+    private int indentationLevel = 0;
+
+    // 2 spaces per indentation
+    public static final String INDENT = "  ";
+
+    // stores N concatenations of INDENT at index N, where index 0 stores empty string
+    // the index is the indentation level
+    protected final List<String> indentStrings = new ArrayList<>();
+
+    // used to format constant slots
+    protected final AnnotationFormatter formatter;
 
     public ToStringSerializer(boolean showAstPaths) {
         this.showAstPaths = showAstPaths;
         this.showVerboseVars = true;
+        formatter = InferenceMain.getInstance().getRealTypeFactory().getAnnotationFormatter();
+
+        // set first value to ""
+        indentStrings.add("");
     }
 
-    public void setIndent(int indent) {
-        this.indent = indent;
+    private void generateIndentations() {
+        // create additional indentation strings up to the current level of indentation, if it
+        // doesn't exist in the indentStrings array list
+        // subsequent indentation string values are 1 more INDENT compared to the previous index
+        StringBuilder sb = new StringBuilder(indentStrings.get(indentStrings.size() - 1));
+        for (int i = indentStrings.size(); i <= getIndentationLevel(); i++) {
+            indentStrings.add(sb.append(INDENT).toString());
+        }
     }
 
-    public int getIndent() {
-        return indent;
+    public void setIndentationLevel(int indentationLevel) {
+        // Ensure the indentation level is always >= 0
+        this.indentationLevel = indentationLevel > 0 ? indentationLevel : 0;
+    }
+
+    public int getIndentationLevel() {
```

Comment:
Can we remove this method? There is no client use this method at all.

Reply:
its used within this class, i'll make it private